src/go/cmd/strelka-frontend/main.go

```diff
@@ -226,6 +238,279 @@ func (s *server) ScanFile(stream strelka.Frontend_ScanFileServer) error {
+	}
+
+	if err := stream.Send(resp); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *server) SyncYara(stream strelka.Frontend_SyncYaraServer) error {
+	var orgID string
+	var req *strelka.Request
+
+	deadline, ok := stream.Context().Deadline()
+	if ok == false {
+		return nil
+	}
+
+	id := uuid.New().String()
+
+	var keyYaraHash string
+	keyOrgID := fmt.Sprintf("org_id:%s", id)
+	keyYaraSync := fmt.Sprintf("yara:compile_and_sync:%s", id)
+	keyYaraSyncDone := fmt.Sprintf("yara:compile_and_sync:done:%s", id)
+
+	for {
+		in, err := stream.Recv()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+
+		if req == nil {
+			req = in.Request
+		}
+
+		if orgID == "" {
+			orgID = in.OrgID
+		}
+
+		p := s.coordinator.cli.Pipeline()
+		keyYaraHash = fmt.Sprintf("yara:hash:%s", orgID)
+		p.Set(stream.Context(), keyOrgID, orgID, time.Until(deadline))
+
+		if len(in.Data) > 0 {
+			for _, inData := range in.Data {
+				outData, err := json.Marshal(*inData)
+				if err != nil {
+					return err
+				}
+
+				// Send for compilation
+				p.RPush(stream.Context(), keyYaraSync, outData)
+				p.ExpireAt(stream.Context(), keyYaraSync, deadline)
+			}
+
+			if _, err := p.Exec(stream.Context()); err != nil {
+				return err
+			}
+		}
+	}
+
+	// skip gatekeeper, we're not sending it
+
+	// send task to backend
+	if err := s.coordinator.cli.ZAdd(
+		stream.Context(),
+		"tasks_compile_and_sync_yara",
+		&redis.Z{
+			Score:  float64(deadline.Unix()),
+			Member: id,
+		},
+	).Err(); err != nil {
+		return err
+	}
+
+	var errMsg string
+
+	for {
+		if err := stream.Context().Err(); err != nil {
+			return fmt.Errorf("context closed: %w", err)
+		}
+
+		res, err := s.coordinator.cli.BLPop(
+			stream.Context(),
+			5*time.Second,
+			keyYaraSyncDone,
+		).Result()
+		if err != nil {
+			if err != redis.Nil {
+				// Delay to prevent fast looping over errors
+				log.Printf("err: %v\n", err)
+				time.Sleep(250 * time.Millisecond)
+			}
+			continue
+		}
+
+		if res[1] == "FIN" {
+			break
+		}
+
+		if strings.HasPrefix(res[1], "ERROR:") {
+			errMsg = strings.Replace(res[1], "ERROR:", "", 1)
+			break
+		}
+	}
+
+	resp := &strelka.SyncYaraResponse{}
+
+	resp.Synced = 0
+	hash, err := s.coordinator.cli.Get(stream.Context(), keyYaraHash).Result()
+	if err != nil {
+		return fmt.Errorf("getting hash: %w", err)
+	}
+
+	synced, err := s.coordinator.cli.Get(stream.Context(), fmt.Sprintf("yara:synced:%s", id)).Result()
+	if err != nil {
+		return fmt.Errorf("getting sync count: %w", err)
+	}
+
+	nSynced, err := strconv.Atoi(synced)
+	if err != nil {
+		// bye bye bye
+		return err
+	}
+
+	resp.Hash = []byte(hash)
+	resp.Error = errMsg
+	resp.Synced = int32(nSynced)
+
+	if err := stream.Send(resp); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *server) ShouldUpdateYara(stream strelka.Frontend_ShouldUpdateYaraServer) error {
+	var keyYaraHash string
+	var orgID string
+	var hash []byte
+
+	for {
+		in, err := stream.Recv()
```

Comment:
Do you know if `stream.Recv()` respects the `stream.Context()`? I imagine so, but if we're not entirely sure we should add:

```
		if err := stream.Context().Err(); err != nil {
			return fmt.Errorf("context closed: %w", err)
		}
```

to ensure we'll eventually break out of the loop. This might apply for other loops too -- we just need to be sure that we have a breakout path for when the context expires.

Reply:
Added just in case.